### ISSUE TITLE
Neural network has too many parameters and uses incorrect loss function

### ISSUE THREAD

**Issue by vojtamolda** [marrrcin/pytorch-resnet-mnist#1]
Hello @marrrcin,

I was googling for a quick-and-dirty piece of PyTorch code doing MNIST classification with ResNet-like architecture and I ended up reading your nice [blog post](https://zablo.net/blog/post/using-resnet-for-mnist-in-pytorch-tutorial). While copying-and-pasting your solution I noticed two problems. I think they're worth noting and fixing.

1. The `MnistResNet` model has about 11.2M parameters. This is way too many for a this tiny dataset. Just for comparison the MNIST testing dataset has about 7.2M (10k * 28 * 28) pixels in total.

    Something much much smaller, like a simple 3 residual block network, should easily get the same performance (i.e. 99% accuracy in less than 5 epochs) with much smaller number of parameters. My gut feeling is that ~100k parameters should be more than enough.

2. The loss function is slightly incorrect. `MnistResNet` network directly outputs the softmax probabilities of each class and not scores which is a more typical choice. To get the correct data likelihood maximization objective in this case you should be using a simple negative log likelihood loss `nn.NLLLoss` of the log-softmax probabilities. Applying `nn.CrossEntropyLoss` on the probabilities vector adds something like an "extra log" to the cost function.

    What you're doing isn't wrong since taking logarithm of the objective doesn't affect the location of the minimum (at least in theory). In practice this leads to smaller gradients and slower training. The optimizer probably partially compensates against it by using a larger learning rate but the setup isn't ideal. See the [docs here](https://pytorch.org/docs/stable/nn.html#torch.nn.CrossEntropyLoss) for more detail.

I can submit a quick pull request to fix these two problems if you want. Just let me know.

Cheers,

  Vojta

**Comment by marrrcin** [comment]
Hi Vojta!
This is a great feedback, thanks for pointing out the sloppy parts! I will be glad to merge your PR.

**Comment by vojtamolda** [comment]
Below is a ResNet-like architecture with about 80k parameters that I put together. It uses only 3x3 convolutions. There's an initial convolution and three ResNet blocks stacked on top of each other. The final layer does a simple average pooling instead of fully connected layer. Please, let me know if you have questions or something doesn't make sense.

As usual feel free to stack more blocks on top of each other to get better performance if you need it.

![MnistResNet](https://user-images.githubusercontent.com/8467740/55291703-47c13900-53a7-11e9-8f47-1486313d465c.png)

The network achieved about 99.3% validation accuracy on MNIST and trained for about 10 mins and 5 epochs on my laptop's CPU.

```python
class MnistResNet(nn.Module):

    class Block(nn.Module):
        def __init__(self, in_channels: int, out_channels: int):
            super().__init__()
            self.conv1 = self.conv3x3(in_channels, out_channels, stride=2)
            self.bn1 = nn.BatchNorm2d(out_channels)
            self.relu = nn.ReLU(inplace=True)
            self.conv2 = self.conv3x3(out_channels, out_channels)
            self.bn2 = nn.BatchNorm2d(out_channels)

            self.convd = nn.Conv2d(in_channels, out_channels, kernel_size=1, stride=2, bias=False)
            self.bnd = nn.BatchNorm2d(out_channels)

        @staticmethod
        def conv3x3(in_channels: int, out_channels: int, stride: int = 1) -> nn.Conv2d:
            return nn.Conv2d(in_channels, out_channels, kernel_size=3, stride=stride, padding=1, bias=False)

        def forward(self, input: torch.Tensor) -> torch.Tensor:
            residual = self.conv1(input)
            residual = self.bn1(residual)
            residual = self.relu(residual)
            residual = self.conv2(residual)
            residual = self.bn2(residual)

            downsample = self.convd(input)
            downsample = self.bnd(downsample)

            out = residual + downsample
            out = self.relu(out)
            return out

    def __init__(self):
        super().__init__()
        self.conv1 = self.Block.conv3x3(1, 16)
        self.bn1 = nn.BatchNorm2d(16)
        self.relu = nn.ReLU(inplace=True)

        self.block1 = self.Block(16, 32)
        self.block2 = self.Block(32, 64)
        self.block3 = self.Block(64, 10)

        self.avgpool = nn.AvgPool2d(kernel_size=28)

    def forward(self, input: torch.Tensor) -> torch.Tensor:
        out = self.conv1(input)
        out = self.bn1(out)
        out = self.relu(out)

        out = self.block1(out)
        out = self.block2(out)
        out = self.block3(out)

        out = self.avgpool(out)
        return out.squeeze()

```

I didn't have time to do a proper PR so I've attached a stand-alone Python script that does the training. I tried to replace the `MnistResNet` class in `pytorch-resnet-mnist.ipynb` notebook but the training was super slow and I couldn't easily understand why. I'm not a big fan of Jupyter notebooks and honestly I don't even know how to put a breakpoint there to debug. So I hope you'll have better luck seeing what I missed.

The script also contains a simple VGG-style architecture for reference and comparison. It achieves very similar accuracy, trains faster but has about 3.5x more parameters.

[`mnist.py`](https://github.com/marrrcin/pytorch-resnet-mnist/files/3026854/mnist.txt)

The exact commad line arguments I used are these:
```bash
python3 mnist.py --network_type=resnet --max_epochs=5 --logdir=tensorboard_log
```
